File: Data Structures/Trees/Binary Tree/BinaryTree.java
import java.util.Queue;
import java.util.LinkedList;

class Node{
    int data;
    Node left;
    Node right;

    Node(int data1){
        data = data1;
        left = null;
        right = null;
    }
};

public class BinaryTree {
    public static Node head = new Node(1);

    public static void insertion(Node  head, int data){
    Node x = new Node(data);
    if(head == null){
        head = x;
        return;
    }
    Queue<Node> q = new LinkedList<>();
    q.add(head);
    while(!q.isEmpty()){
        Node temp = q.peek();
        q.remove();
        if(temp.left == null){
            temp.left = x;
            return;
        }
        else{
            q.add(temp.left);
        }
        if(temp.right == null){
            temp.right = x;
            return;
        }
        else{
            q.add(temp.right);
        }    
    }

}

public static void preorder(Node head){
    if (head==null){
        return;
    }
    System.out.printf("%d,", head.data);
    preorder(head.left);
    preorder(head.right);
}

public static void inorder(Node head){
    if (head==null){
        return;
    }
    inorder(head.left);
    System.out.printf("%d,", head.data);
    inorder(head.right);
}

public static void postorder(Node head){
    if (head==null){
        return;
    }
    postorder(head.left);
    postorder(head.right);
    System.out.printf("%d,", head.data);
}

public static void levelorder(Node head){
    Queue<Node> q = new LinkedList<>();
    q.add(head);
    while(!q.isEmpty()){

        Node temp = q.peek();
        System.out.printf("%d,", temp.data);
        q.remove();

        if(temp.left != null){
            q.add(temp.left);
        }
        
        if(temp.right != null){
            q.add(temp.right);
        }

    }
}


public static void deletionDeep(Node  head, Node  x){
    Queue<Node> q = new LinkedList<>();
    q.add(head);
    while(!q.isEmpty()){
        Node temp = q.peek();
        q.remove();
        if(temp == x){
            temp = null;
            return;
        }
        if(temp.left == x){
            temp.left = null;
            return;
        }
        else{
            q.add(temp.left);
        }
        if(temp.right == x){
            temp.right = null;
            return;
        }
        else{
            q.add(temp.right);
        }    
    }

}

public static void deletion(Node  head, int data){
    Queue<Node> q = new LinkedList<>();
    q.add(head);
    Node temp = head;
    Node keyNode = null;
    while(!q.isEmpty()){
        temp = q.peek();
        q.remove();
        if(temp.data == data){
            keyNode = temp;
        }
        if(temp.left != null){
            q.add(temp.left);
        }
        if(temp.right != null){
            q.add(temp.right);
        }  
    }
    if(keyNode != null){
        int x = temp.data;
        keyNode.data = x;
        deletionDeep(head, temp);
    }
}


public static void main(String args[]){
    
    insertion(head, 2);
    insertion(head, 3);
    insertion(head, 4);
    insertion(head, 5);

    System.out.printf("preorder: ");
    preorder(head);
    System.out.printf("\n");

    System.out.printf("inorder: ");
    inorder(head);
    System.out.printf("\n");

    System.out.printf("postorder: ");
    postorder(head);
    System.out.printf("\n");

    System.out.printf("levelorder: ");
    levelorder(head);
    System.out.printf("\n");

    deletion(head, 3);

    System.out.printf("levelorder: ");
    levelorder(head);
    System.out.printf("\n");

}

}

File: Data Structures/Trees/Binary Tree/BinaryTree.cpp
#include <iostream>
#include<queue>
using namespace std;

class Node{
    public: 
    int data;
    Node* left;
    Node* right;

    public:
    Node(int data1){
        data = data1;
        left = NULL;
        right = NULL;
    }
};

void insertion(Node ** head, int data){
    Node* x = new Node(data);
    if(head == NULL){
        *head = x;
        return;
    }
    queue<Node*> q;
    q.push(*head);
    while(!q.empty()){
        Node* temp = q.front();
        q.pop();
        if(temp->left == NULL){
            temp->left = x;
            return;
        }
        else{
            q.push(temp->left);
        }
        if(temp->right == NULL){
            temp->right = x;
            return;
        }
        else{
            q.push(temp->right);
        }    
    }

}

void preorder(Node* head){
    if (head==NULL){
        return;
    }
    cout << head->data << ",";
    preorder(head->left);
    preorder(head->right);
}

void inorder(Node* head){
    if (head==NULL){
        return;
    }
    inorder(head->left);
    cout << head->data << ",";
    inorder(head->right);
}

void postorder(Node* head){
    if (head==NULL){
        return;
    }
    postorder(head->left);
    postorder(head->right);
    cout << head->data << ",";
}

void levelorder(Node* head){
    queue<Node*> q;
    q.push(head);
    while(!q.empty()){

        Node* temp = q.front();
        cout << temp->data << ",";
        q.pop();

        if(temp->left != NULL){
            q.push(temp->left);
        }
        
        if(temp->right != NULL){
            q.push(temp->right);
        }

    }
}


void deletionDeep(Node * head, Node * x){
    queue<Node*> q;
    q.push(head);
    while(!q.empty()){
        Node* temp = q.front();
        q.pop();
        if(temp == x){
            temp = NULL;
            delete x;
            return;
        }
        if(temp->left == x){
            temp->left = NULL;
            delete x;
            return;
        }
        else{
            q.push(temp->left);
        }
        if(temp->right == x){
            temp->right = NULL;
            delete x;
            return;
        }
        else{
            q.push(temp->right);
        }    
    }

}

void deletion(Node * head, int data){
    queue<Node*> q;
    q.push(head);
    Node* temp;
    Node* keyNode = NULL;
    while(!q.empty()){
        temp = q.front();
        q.pop();
        if(temp->data == data){
            keyNode = temp;
        }
        if(temp->left != NULL){
            q.push(temp->left);
        }
        if(temp->right != NULL){
            q.push(temp->right);
        }  
    }
    if(keyNode != NULL){
        int x = temp->data;
        keyNode->data = x;
        deletionDeep(head, temp);
    }
}


int main(){
    Node* head = new Node(1);
    insertion(&head, 2);
    insertion(&head, 3);
    insertion(&head, 4);
    insertion(&head, 5);

    cout <<"preorder: ";
    preorder(head);
    cout << endl;

    cout <<"inorder: ";
    inorder(head);
    cout << endl;

    cout <<"postorder: ";
    postorder(head);
    cout << endl;

    cout <<"levelorder: ";
    levelorder(head);
    cout << endl;

    deletion(head, 3);

    cout <<"levelorder: ";
    levelorder(head);
    cout << endl;
    return 0;


}
